x/qgb/abci_test.go

```diff
@@ -271,3 +277,132 @@ func TestDataCommitmentRange(t *testing.T) {
+				require.NoError(t, err)
+			},
+			expectExists: false,
+		},
+		{
+			name: "when store has one data commitment",
+			setup: func() {
+				ctx = ctx.WithBlockHeight(400)
+				dc, err := qk.NextDataCommitment(ctx)
+				require.NoError(t, err)
+				err = qk.SetAttestationRequest(ctx, &dc)
+				require.NoError(t, err)
+			},
+			expectExists: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt.setup()
+			exists, err := qk.HasDataCommitmentInStore(ctx)
+			require.NoError(t, err)
+			assert.Equal(t, tt.expectExists, exists)
+		})
+	}
+}
+
+// TestGetDataCommitment This test will test the data commitment creation catchup mechanism.
+// It will run `abci.EndBlocker` on all the heights while changing the data commitment window
+// in different occasions, to see if at the end of the test, the data commitments cover all
+// the needed ranges.
+func TestDataCommitmentCreationCatchup(t *testing.T) {
+	input, ctx := testutil.SetupFiveValChain(t)
+	qk := input.QgbKeeper
+	ctx = ctx.WithBlockHeight(1)
+
+	executeHeights := func(beginHeight int64, endHeight int64) {
+		for i := beginHeight; i <= endHeight; i++ {
+			ctx = ctx.WithBlockHeight(i)
+			qgb.EndBlocker(ctx, *qk)
+		}
+	}
+
+	// from height 1 to 1500 with a window of 400
+	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 400})
+	executeHeights(1, 1500)
+
+	// change window to 100 and execute up to 1920
+	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 100})
+	executeHeights(1501, 1920)
+
+	// change window to 1000 and execute up to 3500
+	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 1000})
+	executeHeights(1921, 3500)
+
+	// change window to 111 and execute up to 3800
+	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 111})
+	executeHeights(3501, 3800)
+
+	// check if a data commitment was created
+	hasDataCommitment, err := qk.HasDataCommitmentInStore(ctx)
+	require.NoError(t, err)
+	require.True(t, hasDataCommitment)
+
+	// get the last attestation nonce
+	lastAttestationNonce := qk.GetLatestAttestationNonce(ctx)
+
+	// check if the ranges are contineous
+	var previousDC types.DataCommitment
+	dcCount := 0
+	for i := uint64(1); i <= lastAttestationNonce; i++ {
+		att, found, err := qk.GetAttestationByNonce(ctx, i)
+		require.NoError(t, err)
+		require.True(t, found)
+		dc, ok := att.(*types.DataCommitment)
+		if !ok {
+			continue
+		}
+		dcCount++
+		if previousDC.Nonce == 0 {
+			// initialize the previous dc
+			previousDC = *dc
+			continue
+		}
+		assert.Equal(t, previousDC.EndBlock+1, dc.BeginBlock)
+		previousDC = *dc
+	}
+
+	// we should have 19 data commitments created in the above setup
+	// - window 400: [1, 400], [401, 800], [801, 1200]
+	// - window 100: [1201, 1300], [1301, 1400], [1401, 1500], [1501, 1600], [1601, 1700], [1701, 1800], [1801, 1900]
+	// - window 1000: [1901, 2900]
+	// - window 111: [2901, 3011], [3012, 3122], [3123,3233], [3234, 3344], [3345, 3455], [3456, 3566], [3567, 3677], [3678, 3788]
```

Comment:
[optional] this comment is soooo helpful so I converted it into an explicit `want`. Totally optional:

```go
// TestGetDataCommitment This test will test the data commitment creation catchup mechanism.
// It will run `abci.EndBlocker` on all the heights while changing the data commitment window
// in different occasions, to see if at the end of the test, the data commitments cover all
// the needed ranges.
func TestDataCommitmentCreationCatchup(t *testing.T) {
	input, ctx := testutil.SetupFiveValChain(t)
	qk := input.QgbKeeper
	ctx = ctx.WithBlockHeight(1)

	executeHeights := func(beginHeight int64, endHeight int64) {
		for i := beginHeight; i <= endHeight; i++ {
			ctx = ctx.WithBlockHeight(i)
			qgb.EndBlocker(ctx, *qk)
		}
	}

	// from height 1 to 1500 with a window of 400
	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 400})
	executeHeights(1, 1500)

	// change window to 100 and execute up to 1920
	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 100})
	executeHeights(1501, 1920)

	// change window to 1000 and execute up to 3500
	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 1000})
	executeHeights(1921, 3500)

	// change window to 111 and execute up to 3800
	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 111})
	executeHeights(3501, 3800)

	// check if a data commitment was created
	hasDataCommitment, err := qk.HasDataCommitmentInStore(ctx)
	require.NoError(t, err)
	require.True(t, hasDataCommitment)

	// get the last attestation nonce
	lastAttestationNonce := qk.GetLatestAttestationNonce(ctx)

	got := []types.DataCommitment{}
	for i := uint64(1); i <= lastAttestationNonce; i++ {
		att, found, err := qk.GetAttestationByNonce(ctx, i)
		require.NoError(t, err)
		require.True(t, found)
		dc, ok := att.(*types.DataCommitment)
		if !ok {
			continue
		}
		got = append(got, *dc)
	}

	// We should have 19 data commitments created in the above setup
	// - window 400: [1, 400], [401, 800], [801, 1200]
	// - window 100: [1201, 1300], [1301, 1400], [1401, 1500], [1501, 1600], [1601, 1700], [1701, 1800], [1801, 1900]
	// - window 1000: [1901, 2900]
	// - window 111: [2901, 3011], [3012, 3122], [3123,3233], [3234, 3344], [3345, 3455], [3456, 3566], [3567, 3677], [3678, 3788]
	want := []types.DataCommitment{
		{
			Nonce:      2, // nonce 1 is the valset attestation
			BeginBlock: 1,
			EndBlock:   400,
		},
		{
			Nonce:      3,
			BeginBlock: 401,
			EndBlock:   800,
		},
		{
			Nonce:      4,
			BeginBlock: 801,
			EndBlock:   1200,
		},
		{
			Nonce:      5,
			BeginBlock: 1201,
			EndBlock:   1300,
		},
		{
			Nonce:      6,
			BeginBlock: 1301,
			EndBlock:   1400,
		},
		{
			Nonce:      7,
			BeginBlock: 1401,
			EndBlock:   1500,
		},
		{
			Nonce:      8,
			BeginBlock: 1501,
			EndBlock:   1600,
		},
		{
			Nonce:      9,
			BeginBlock: 1601,
			EndBlock:   1700,
		},
		{
			Nonce:      10,
			BeginBlock: 1701,
			EndBlock:   1800,
		},
		{
			Nonce:      11,
			BeginBlock: 1801,
			EndBlock:   1900,
		},
		{
			Nonce:      12,
			BeginBlock: 1901,
			EndBlock:   2900,
		},
		{
			Nonce:      13,
			BeginBlock: 2901,
			EndBlock:   3011,
		},
		{
			Nonce:      14,
			BeginBlock: 3012,
			EndBlock:   3122,
		},
		{
			Nonce:      15,
			BeginBlock: 3123,
			EndBlock:   3233,
		},
		{
			Nonce:      16,
			BeginBlock: 3234,
			EndBlock:   3344,
		},
		{
			Nonce:      17,
			BeginBlock: 3345,
			EndBlock:   3455,
		},
		{
			Nonce:      18,
			BeginBlock: 3456,
			EndBlock:   3566,
		},
		{
			Nonce:      19,
			BeginBlock: 3567,
			EndBlock:   3677,
		},
		{
			Nonce:      20,
			BeginBlock: 3678,
			EndBlock:   3788,
		},
	}
	assert.Equal(t, 19, len(got))
	assert.Equal(t, want, got)
}
```